### DISCUSSION TITLE
怎样在RailWay部署？

### DISCUSSION THREAD

**Discussion by yingyuezhiyun** [swuecho/chat#112] [Q&A] [answered]
不是很明白docker原理，想请教如何在railway上进行部署

**Comment by swuecho** [comment]
0. 准备项目

回到 cmd line:
```
git clone git@github.com:swuecho/chat.git
# build the frontend code
cd web;
pnpm install
pnpm build
cd ..; cp -R web/dist/* api/static/
cd api
```

1. install railway cmd line.

https://docs.railway.app/develop/cli
```
railway login
```

**Comment by swuecho** [comment]
3. 
```railway init```

will init a new project, go to that project

<img width="1701" alt="image" src="https://user-images.githubusercontent.com/666683/232217742-2563d82a-2596-47f0-ba64-f3f89c04d8a9.png">

**Comment by swuecho** [comment]
add  service, select postgresql


<img width="1513" alt="image" src="https://user-images.githubusercontent.com/666683/232218154-6e370680-4bc7-4f38-b0b3-ef40cd2fc052.png">

**Comment by swuecho** [comment]
railway 创建的database 会自动 提供一个 DATABASE_URL 的环境变量.

**Comment by swuecho** [comment]
click the "settings", add env var, api key 没有的乱填值就行了.

<img width="1298" alt="image" src="https://user-images.githubusercontent.com/666683/232219330-5127f733-2f10-4326-9efe-7cac17fb24ec.png">

**Comment by swuecho** [comment]
railway link
> Select a project (the project you init above)
> Select an environment production

**Comment by swuecho** [comment]
然后, railway up, 需要在 chat/api 目录下

**Comment by swuecho** [comment]
回到 railway 网页, 添加var

<img width="1124" alt="image" src="https://user-images.githubusercontent.com/666683/232221013-b49122c6-869b-4036-9341-acb44cd35725.png">

**Comment by swuecho** [comment]
需要一下五个:
PORT=8080 

因为main.go 里面写的是8080

<img width="1152" alt="image" src="https://user-images.githubusercontent.com/666683/232222694-8920281f-8d78-46aa-aeeb-696d081c864e.png">

**Comment by swuecho** [comment]
然后生成域名:

<img width="1049" alt="image" src="https://user-images.githubusercontent.com/666683/232221441-5f0ec558-5eeb-410a-a914-838bbaad8763.png">

**Comment by swuecho** [comment]
应该就可以了.

我已经成功部署并chat 了.

**Comment by swuecho** [comment]
生成template 倒也是可以, 因为web, api 在 独立的folder 我得重写 Dockerfile, 有空再看吧.

**Comment by swuecho** [comment]
部署之后,  注册用户, 第一个用户是管理员, 然后到  https://$hostname/static/#/admin/user 设置 ratelimit, 公网部署, 只对信任的email 增加 ratelimit, 这样即使有人注册, 也是不能用的. 

<img width="1851" alt="image" src="https://user-images.githubusercontent.com/666683/232227529-284289a8-1336-49dd-b5c6-8e8226b9e862.png">

**Comment by yingyuezhiyun** [comment]
非常感谢您的耐心解答，非常详细的步骤！我明天去试验一下。

**Comment by yingyuezhiyun** [comment]
非常感谢您的耐心解答，非常详细的步骤！我明天去试验一下。

**Comment by swuecho** [comment, marked as answer]
用这个template

[![Deploy on Railway](https://railway.app/button.svg)](https://railway.app/template/tk7jWU?referralCode=5DMfQv)

然后配置环境变量就可以了.

PORT=8080
RATELIMITE=0

别的两个key 有就填.

<img width="1155" alt="image" src="https://user-images.githubusercontent.com/666683/232234418-941c9336-783c-4430-857c-9e7b703bb1c1.png">

**Comment by swuecho** [comment]
railway 感觉要超过 fly.io 啊, 感觉丝滑很多.

**Comment by yingyuezhiyun** [comment]
非常感谢，大佬真效率！
railway部署一些小项目还是挺好的，只是每个月有500小时限制，绑卡才能解锁。

**Comment by yingyuezhiyun** [comment]
再次感谢，我已经成功在railway上部署.
请问大佬后续有考虑加入OPENAI_ACCESS_TOKEN吗？现在申请账号已经没免费额度了😂，visa信用卡学生党又很难申请🙃

**Comment by swuecho** [comment]
OPENAI_ACCESS_TOKEN

这个不是很懂, 看看有没有懂的. 

你可以申请一个Claude的 early access key 目前是免费的,我这个项目也支持他家的key, 效果目前感觉快跟上open ai的模型了, 就是通过慢一点.

**Comment by swuecho** [comment]
如果你自己用的话, 可以在 demo project 上面 https://chat-production-d90e.up.railway.app/ 注册个号, 我给你加ratelimit 你可以用, 这个demo project 我每次更新代码都会更新, 可能会有bug (顺便帮测试). 

我暂时准备放弃自己本地部署了, 没有railway 这个自动更新方便.

**Comment by yingyuezhiyun** [comment]
> OPENAI_ACCESS_TOKEN
> 
> 这个不是很懂, 看看有没有懂的.
> 
> 你可以申请一个Claude的 early access key 目前是免费的,我这个项目也支持他家的key, 效果目前感觉快跟上open ai的模型了, 就是通过慢一点.

👌，我前几天申请了Claude API KEY目前还没收到反馈，再等几天看看

**Comment by yingyuezhiyun** [comment]
> 如果你自己用的话, 可以在 demo project 上面 https://chat-production-d90e.up.railway.app/ 注册个号, 我给你加ratelimit 你可以用, 这个demo project 我每次更新代码都会更新, 可能会有bug (顺便帮测试).
> 
> 我暂时准备放弃自己本地部署了, 没有railway 这个自动更新方便.

好的，非常感谢，多设备同步消息真的挺方便

**Reply by swuecho** [reply]
20提问/10min 应该够了
